src/rng.rs

```diff
@@ -1,28 +1,58 @@
 use crate::Vec;
-use algebra::{FromBytes, ToBytes, Field};
+use algebra::{FromBytes, ToBytes, to_bytes, Field};
 use std::{
     convert::TryInto,
-    marker::PhantomData
+    marker::PhantomData,
 };
 use digest::{generic_array::GenericArray, Digest};
 use rand_chacha::ChaChaRng;
 use rand_core::{RngCore, SeedableRng};
 use rand::Rng;
+use crate::error::Error;
+
+/// General trait for an updatable seed to be used to initialize a `FiatShamirRng`.
+pub trait FiatShamirRngSeed {
+    /// Output type of the seed, to be consistent with the seed type accepted
+    /// by the `from_seed` function of FiatShamirRng.
+    type FinalizedSeed: Sized + Clone;
+
+    /// Error type
+    type Error: std::error::Error + From<Error>;
+
+    /// Maximum number of elements allowed to be part of this seed
+    const MAX_NUM_ELEMENTS: usize;
+
+    /// Maximum length of each element that is part of the seed
+    const MAX_ELEMENT_LEN: usize;
+
+    /// Initialize this seed
+    fn new() -> Self;
+
+    /// Update this seed with a new element interpreted as a sequence of byte
+    fn add_bytes<'a, T: 'a + ToBytes>(&mut self, elem: &'a T) -> Result<&mut Self, Self::Error>;
+
+    /// Finalize this seed to the type needed by the corresponding FiatShamirRng impl
+    fn finalize(self) -> Self::FinalizedSeed;
+}
```

Comment:
Is a separate trait for the seed really necessary?
All we need is a "true" serialization routine (which is one-to-one as every usual serialization). So why not implementing CanonicalSerialize for the type of seeds in question?

Reply:
I wanted to be more generic and I found this to be the preferred solution. 
Your suggestion means the necessity to introduce a new struct for each different type of seed, implement CanonicalSerialize for it and use it only in the specific context without the possibility to call it in an abstract way.

Reply:
Hm... not sure if I understand. Isn't a seed comprised of elements for which we typically have an implementation of CanonicalSerialize?